File: 365.water-and-jug-problem.py
#
# @lc app=leetcode id=365 lang=python3
#
# [365] Water and Jug Problem
#

# Time:  O(logn)
# Space: O(1)

# @lc code=start
class Solution:
    def canMeasureWater(self, x: int, y: int, z: int) -> bool:
        if z == 0:
            return True
        elif x + y < z:
            return False

        def gcd(a, b):
            while b:
                a, b = b, a % b
            return a

        return z % gcd(x, y) == 0


# @lc code=end

File: 365.water-and-jug-problem.js
/*
 * @lc app=leetcode id=365 lang=javascript
 *
 * [365] Water and Jug Problem
 */

// @lc code=start
/**
 * @param {number} x
 * @param {number} y
 * @param {number} z
 * @return {boolean}
 */
var canMeasureWater = function (x, y, z) {
  const gcd = (a, b) => {
    return b === 0 ? a : gcd(b, a % b);
  };
  if (z === 0) {
    return true;
  } else if (x + y >= z && z % gcd(x, y) === 0) {
    return true;
  }
  return false;
};
// @lc code=end
